File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

//This script checks inputted strings for validity as credit card numbers
int getLength(long long input){
    int length = 0;
    while(input > 0){
        length++;
        input /= 10;
    }
    return length;
}

string getCardType(long long input, int length){
    if(length == 15){
        long long firstTwo = input / 10000000000000;
        if(firstTwo == 37 || firstTwo == 34)
            return "AMEX\n";
    }
    else if(length == 16){
        long long firstTwo = input / 100000000000000;
        if(firstTwo == 51 || firstTwo == 52 || firstTwo == 53 || firstTwo == 54 || firstTwo == 55)
            return "MASTERCARD\n";
        else{
            firstTwo /= 10;
            if(firstTwo == 4){
                return "VISA\n";
            }
        } 
    }
    else{
        long long firstTwo = input / 1000000000000;
        if(firstTwo == 4){
            return "VISA\n";
        }
    
    }
    return "INVALID";
}


bool checkValidity(long long input){
    if(input < 0)
        return false;
    int length = getLength(input);
    
    if(length == 13 || length == 15 || length == 16){
        int sumOfTimes = 0;
        int sumOthers = 0;
        int sumOfTimes2 = 0;
        long long temp = input;
        long long temp3 = input;
        while(temp != 0){
            long long temp2 = ((temp % 100) / 10);
            temp /= 100;
            sumOfTimes += (temp2 * 2);
            for(int i = (temp2 * 2); i > 0; i/=10){
                sumOfTimes2 += i % 10;
            }
            
        }
        
        // sum of odd place numbers
        while(temp3 > 0){
            long long temp2 = ((temp3 % 100) % 10);
            temp3 /= 100;
            sumOthers += temp2;
        }
        
        if((sumOfTimes2 + sumOthers) % 10 == 0){
            return true;
        }
    }
    return false;
}



int main(void)
{
    printf("Number: ");
    long input = get_long_long();
    if (checkValidity(input)){
        int length = getLength(input);
        string cardType = getCardType(input, length);
        printf("%s", cardType);
    }
    else{
        printf("INVALID\n");
    }
    
    return 0;
}

File: credit.py
def main():
    number = "0"

    while (len(number) != 13 and len(number) != 15 and len(number) != 16):
        number = input("Number: ")
    length = len(number)
    number = int(number)
    if(checkValidity(number)):
        cardType = getCardType(number, length)
        print(cardType)
    else:
        print("INVALID")


def checkValidity(number):
    sumofTimes = 0
    sumOthers = 0
    temp = number
    temp3 = number
    while(temp != 0):
        temp2 = ((temp % 100) // 10)
        temp //= 100
        temp2 *= 2
        i = temp2
        while(i != 0):
            sumofTimes += i % 10
            i//=10
    while(temp3 != 0):
        temp2 = ((temp3 % 100) % 10)
        temp3 //= 100
        sumOthers += temp2
    if((sumOthers + sumofTimes) % 10 == 0):
        return True
    return False

def getCardType(number, length):
    if(length == 15):
        firstTwo = number // 10000000000000
        if(firstTwo == 37 or firstTwo == 34):
            return "AMEX"
        elif(length == 16):
            firstTwo = number // 100000000000000
            if(firstTwo == 51 or firstTwo == 52 or firstTwo == 53 or firstTwo == 54 or firstTwo == 55):
                return "MASTERCARD"
            else:
                firstTwo //= 10
                if(firstTwo == 4):
                    return "VISA"
        else:
            firstTwo = number // 1000000000000
            if(firstTwo == 4):
                return "VISA"
    return "INVALID"

if __name__ == "__main__":
    main()